lib/Doctrine/DBAL/Logging/NullLogger.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Doctrine\DBAL\Logging;
+
+/**
+ * A SQL logger that does nothing.
+ */
+class NullLogger implements SQLLogger
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function startQuery($sql, array $params = null, array $types = null)
```

Comment:
`?array $params`
`?array $types`

Reply:
This requires changing the interface. And quite a lot of code ([1](https://github.com/doctrine/dbal/blob/develop/lib/Doctrine/DBAL/Connection.php#L1007), [2](https://github.com/doctrine/dbal/blob/develop/lib/Doctrine/DBAL/Connection.php#L1089), [3](https://github.com/doctrine/dbal/blob/develop/lib/Doctrine/DBAL/Connection.php#L1247), ...) only provides the first parameter, this would force them to call `startQuery($query, null, null)` which IMHO is not a very good idea.

Reply:
No it doesn't (I am not suggesting to drop `= null`, only add `?` - also see CS violation on Travis). Nullable is implied by the interface specifying `null` as default value.
https://3v4l.org/mKITu

Reply:
IMO there's no need for params and types to be nullable. We can default them to empty arrays and avoid dealing with nulls: `array $params = [], array $types = []`.

Otherwise, the nullable signature will allow calls like `startQuery('SQL', null, null)` which doesn't make any sense.

Reply:
That is better, but a BC break.

Reply:
True. Still worth doing separately for develop.

Reply:
True. Still worth doing separately for develop.

Reply:
Agreed, let's change the interface then.

Reply:
Done, and because we're changing the interface anyway, I also added `string` type hint and `void` return types.

Reply:
> because we're changing the interface anyway

Well, adding type for parameter is not a BC break (we require PHP 7.2), but adding void is. :) But should be okay for develop. Please document it in UPGRADE.md.

Reply:
Now is the time for BC breaks! I added a note to `UPGRADE.md`.